gamemodes/terrortown/entities/entities/ttt_smoke.lua

```diff
@@ -0,0 +1,396 @@
+	self.Emitter:SetPos(self:GetPos())
+	local c = self:GetSmokeColor():ToColor()
+	local diameter = self:GetRadius() * 2
+
+	for i = 1, self:GetDensity() do
+		local start_size = math.Clamp(RandomWrap(
+			self:GetStartSizeBase() - self:GetStartSizeVariance(),
+			self:GetStartSizeBase() + self:GetStartSizeVariance()
+		), 1, diameter)
+		local max_range = diameter - (start_size / 2)
+		local offset_ratio = (max_range / diameter)
+		local offset_dimension = self:GetRadius() * offset_ratio
+		local rel_pos = self:PointInSphere() * offset_dimension
+		local sprite_dex = math.max(1,math.ceil(RandomWrap(0, #self.SmokeParticles)))
+		local p = self.Emitter:Add(self.SmokeParticles[sprite_dex], self:GetPos() + rel_pos)
+
+		self.particles[ #self.particles + 1 ] = p
+
+		if p then
+			p:SetColor(c.r, c.g, c.b)
+			p:SetStartAlpha(255)
+			p:SetEndAlpha(255)
+
+			p:SetVelocity((p:GetPos() - self:GetPos()):GetNormal() * self.driftVelocity)
+
+			p:SetLifeTime(0)
+			p:SetDieTime(self:GetLifeTime())
+
+			p:SetStartSize(start_size)
+			p:SetEndSize(start_size)
+
+			p:SetRoll(RandomWrap(-self.rollVariance, self.rollVariance))
+			p:SetRollDelta(RandomWrap(-self.rollDeltaVariance, self.rollDeltaVariance))
+
+			p:SetCollide(false)
+			p:SetLighting(false)
+			p:SetBounce(0)
+
+			-- bloom the particles out for their last X seconds of life
+			if self:GetFadeTime() > 0 then
+				self.particleData[ #self.particleData + 1 ] = {
+					end_size = math.max(RandomWrap(
+						self:GetEndSizeBase() - self:GetEndSizeVariance(),
+						self:GetEndSizeBase() + self:GetEndSizeVariance()
+					), 0),
+					future_die_time = RandomWrap(
+						0,
+						self:GetFadeTime()
+					)
+				}
+			end
+		end
+	end
+
+	self.Emitter:Finish()
+end
+
+---
+-- Takes a point in space and blows all our particles away from it, disabling the smoke effect.
+-- Moves particles on client, disales on server.
+-- @param Vector pos The point of origin for the effect that will disperse this smoke.
+-- @param number radius The radius of effect for the thing that is dispersing this smoke.
+-- @param number force How strong the effect of the thing that is dispersing this smoke is.
+-- @realm shared
+function ENT:DisperseFromPoint(pos, radius, force)
+	self:Disipate()
+
+	if not CLIENT then return end
+
+	local distSqr = radius * radius
+
+	for i = 1, self:GetDensity() do
+		local particle = self.particles[i]
+		local dist = particle:GetPos():DistToSqr(pos)
+
+		if dist > distSqr then continue end
+
+		local scale = 1 - (dist / distSqr)
+		local dir = (particle:GetPos() - pos):GetNormal()
+		local pushAngle = dir * force * scale
+
+		particle:SetAirResistance(self.airResistance * self.disperseAirResistMultiplier)
+		particle:SetGravity(-physenv.GetGravity() * self.disperseGravityMultiplier)
+		particle:SetVelocity(particle:GetVelocity() + pushAngle)
+	end
+end
+
+---
+-- Allows all smoke particles to die, and disables the active extinguishing effect.
+-- @realm shared
+function ENT:Disipate()
```

Comment:
AFAIK Dissipate is spelled with two 's'

Reply:
great catch, this is where a JetBrains product would have caught it for me; annoyed that the spellcheck vscode plugin i've got is disregarding code contexts :S